source/common/upstream/upstream_impl.cc

```diff
@@ -1004,13 +1004,22 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(
 
   // Remove hosts from current_priority_hosts that were matched to an existing host in the previous
   // loop.
-  current_priority_hosts.erase(std::remove_if(current_priority_hosts.begin(),
-                                              current_priority_hosts.end(),
-                                              [&existing_hosts_for_current_priority](auto host) {
-                                                return existing_hosts_for_current_priority.count(
-                                                    host->address()->asString());
-                                              }),
-                               current_priority_hosts.end());
+  for (auto itr = current_priority_hosts.begin(); itr != current_priority_hosts.end();) {
+    auto existing_itr = existing_hosts_for_current_priority.find((*itr)->address()->asString());
+
+    if (existing_itr != existing_hosts_for_current_priority.end()) {
+      existing_hosts_for_current_priority.erase(existing_itr);
+      current_priority_hosts.erase(itr);
```

Comment:
Why do we need to erase from `current_priority_hosts`?

Reply:
In this function we loop over all the new hosts (hosts in the config update) and attempt to match them against existing hosts in `all_hosts` (map of address -> host for all hosts for the cluster prior to this update). We keep track of the ones we find in `existing_hosts_for_current_priority` and in this loop we remove the hosts that we've already accounted for from `current_priority_hosts`. The remaining hosts in `current_priority_hosts` are ones that have been removed from this priority. Note that this bit didn't change in this PR, I've only added the part where we set `hosts_changed`.

Thinking about it some, it might be simpler if we pass `current_priority_hosts` as an `unordered_map` (we already make a copy, so converting probably won't be that bad?) and then delete from the map inline instead of keeping track of `existing_hosts_for_current_priority`. Does that sound reasonable?

Reply:
It would also get rid of all these O(n) deletions, which would be nice

Reply:
Looking into implementing my suggestion it's a bit more involved to get right than I originally thought (this part of the code is still pretty complicated).

Might make more sense to get this merged to fix the bug, then revisit it at a later time to refactor if it's useful.